src/reader-activation-auth/google-oauth.js

```diff
@@ -14,38 +26,73 @@ domReady( function () {
 	googleLoginElements.forEach( googleLoginElement => {
 		const googleLoginForm = googleLoginElement.closest( 'form' );
 		const checkLoginStatus = metadata => {
+			debugLog( 'log', '[Google OAuth] checkLoginStatus called with metadata:', metadata );
 			fetch( `/wp-json/newspack/v1/login/google/register?metadata=${ JSON.stringify( metadata ) }` )
 				.then( res => {
+					debugLog( 'log', '[Google OAuth] Register response status:', res.status );
 					res.json()
 						.then( ( { message, data } ) => {
+							debugLog( 'log', '[Google OAuth] Register response data:', { message, data } );
 							if ( googleLoginForm?.endLoginFlow ) {
 								googleLoginForm.endLoginFlow( message, res.status, data );
 							}
 						} )
 						.catch( error => {
+							debugLog( 'error', '[Google OAuth] Error parsing register response:', error );
 							if ( googleLoginForm?.endLoginFlow ) {
 								googleLoginForm.endLoginFlow( error?.message, res.status );
 							}
 						} );
 				} )
 				.catch( error => {
+					debugLog( 'error', '[Google OAuth] Error fetching register endpoint:', error );
 					if ( googleLoginForm?.endLoginFlow ) {
 						googleLoginForm.endLoginFlow( error?.message );
 					}
 				} );
 		};
 
 		googleLoginElement.addEventListener( 'click', () => {
+			debugLog( 'log', '[Google OAuth] Login button clicked' );
 			if ( googleLoginForm?.startLoginFlow ) {
 				googleLoginForm.startLoginFlow();
 			}
 
 			const metadata = googleLoginForm ? convertFormDataToObject( new FormData( googleLoginForm ), [ 'lists[]' ] ) : {};
 			metadata.current_page_url = window.location.href;
+			debugLog( 'log', '[Google OAuth] Opening auth window with metadata:', metadata );
 			const authWindow = window.open( 'about:blank', 'newspack_google_login', 'width=500,height=600' );
-			let googleOAuthSuccess = false;
+			let messageListener = null;
+
+			// Clean up function to remove listeners.
+			const cleanup = () => {
+				debugLog( 'log', '[Google OAuth] Cleanup called' );
+				if ( messageListener ) {
+					window.removeEventListener( 'message', messageListener );
+					messageListener = null;
+				}
+			};
+
+			// Listen for postMessage from OAuth callback.
+			messageListener = event => {
+				debugLog( 'log', '[Google OAuth] Received postMessage:', {
+					origin: event.origin,
+					data: event.data,
+					expectedOrigin: window.location.origin,
+				} );
+				// Validate the message origin matches our domain.
+				if ( event.origin === window.location.origin && event.data === 'google-oauth-success' ) {
+					debugLog( 'log', '[Google OAuth] Valid success message received, calling checkLoginStatus' );
+					cleanup();
+					checkLoginStatus( metadata );
+				}
+			};
+			window.addEventListener( 'message', messageListener );
+
+			// Keep the legacy event listener for backwards compatibility.
 			window.addEventListener( 'google-oauth-success', () => {
```

Comment:
What would still be dispatching this event?

Reply:
Yeah, that's overeager. It was meant as a super-safe backwards compatibility, but there's no need.